lib/block-supports/descendent-block-styles.php

```diff
@@ -0,0 +1,115 @@
+	if ( ! block_has_support( $block_type, array( '__experimentalDescendentStyles' ), false ) ) {
+		return $block_content;
+	}
+
+	// return early if no descendent blocks are found on the block attributes.
+	$block_styles = _wp_array_get( $block, array( 'attrs', 'style', 'blocks' ), null );
+	if ( empty( $block_styles ) ) {
+		return $block_content;
+	}
+
+	$class_name = gutenberg_get_block_styles_class_name( $block );
+
+	// Like the layout hook this assumes the hook only applies to blocks with a single wrapper.
+	// Retrieve the opening tag of the first HTML element.
+	$html_element_matches = array();
+	preg_match( '/<[^>]+>/', $block_content, $html_element_matches, PREG_OFFSET_CAPTURE );
+	$first_element = $html_element_matches[0][0];
+	// If the first HTML element has a class attribute just add the new class
+	// as we do on layout and duotone.
+	if ( strpos( $first_element, 'class="' ) !== false ) {
+		$content = preg_replace(
+			'/' . preg_quote( 'class="', '/' ) . '/',
+			'class="' . $class_name . ' ',
+			$block_content,
+			1
+		);
+	} else {
+		// If the first HTML element has no class attribute we should inject the attribute before the attribute at the end.
+		$first_element_offset = $html_element_matches[0][1];
+		$content              = substr_replace( $block_content, ' class="' . $class_name . '"', $first_element_offset + strlen( $first_element ) - 1, 0 );
+	}
+
+	return $content;
+}
+
+/**
+ * Render the descendent block styles stylesheet.
+ *
+ * @param string|null $pre_render   The pre-rendered content. Default null.
+ * @param array       $block The block being rendered.
+ *
+ * @return null
+ */
+function gutenberg_render_descendent_block_styles_support_styles( $pre_render, $block ) {
+	// Return early if the block has not support for descendent block styles.
+	$block_type           = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	if ( ! block_has_support( $block_type, array( '__experimentalDescendentStyles' ), false ) ) {
+		return null;
+	}
+
+	// Return early if no descendent blocks are found on the block attributes.
+	$block_styles = _wp_array_get( $block, array( 'attrs', 'style', 'blocks' ), null );
+	if ( empty( $block_styles ) ) {
+		return null;
+	}
+
+	$class_name = gutenberg_get_block_styles_class_name( $block );
+
+	// Remove any potentially unsafe styles.
+	$theme_json_shape = WP_Theme_JSON_Gutenberg::remove_insecure_properties( array(
+		'version' =>  WP_Theme_JSON_Gutenberg::LATEST_SCHEMA,
+		'styles' => array(
+			'blocks' => $block_styles,
+		),
+	) );
+
+	$styles = '';
+
+	$theme_json_object = new WP_Theme_JSON_Gutenberg( $theme_json_shape );
+	$block_nodes       = $theme_json_object->get_styles_block_nodes();
+	foreach ( $block_nodes as $block_node ) {
+		$block_node['selector'] = WP_Theme_JSON_Gutenberg::scope_selector( '.' . $class_name, $block_node['selector'] );
+		$styles .= $theme_json_object->get_styles_for_block( $block_node );
```

Comment:
Is there another way we can output these styles? The pattern of using id-based classnames and outputting additional `style` tags is something we're attempting to move away from in the style engine (https://github.com/WordPress/gutenberg/issues/38167) and layout projects (https://github.com/WordPress/gutenberg/issues/39336). Ideally it'd be great if we can reduce the amount we're relying on these approaches, as they're not covered by rendering in async contexts (e.g. the problems outlined in  https://github.com/WordPress/gutenberg/issues/35376)

I was wondering if it's possible for the styles for descendent blocks to somehow be output from the global styles sheet? I'm not too familiar with the work for this block support, but the idea of having the required styles available in the global style sheet is one of the things I'm exploring in the Layout block support refactor in: https://github.com/WordPress/gutenberg/pull/40875

I'll also CC: @ramonjd for visibility here, since we've been discussing some of these concerns for future phases of the style engine lately 🙂

Reply:
> The pattern of using id-based classnames and outputting additional style tags is something we're attempting to move away from in the style engine

Thanks @andrewserong 

It's not just about the style engine, but the long-standing issues we've been task to address with it. Also refer to https://make.wordpress.org/core/2022/06/24/block-editor-styles-initiatives-and-goals/ for background.

Two of them are to rein in the specificity issues caused by random IDs and to construct more semantic classnames that theme authors can rely on. 

Are the unique ids required to target elements with custom CSS, unique to that element? Or for specificity?

Maybe there's not a better way to achieve what you're trying to do right now, but it'd be useful to see if we can come up with an alternative that won't compound the style quality issues theme users already dealing with.

Reply:
> I was wondering if it's possible for the styles for descendent blocks to somehow be output from the global styles sheet? 

I'm also wondering if theme.json/global styles is a logical home for this behaviour. 

How will this interact with the current block supports mechanism? 

If this data lives in the block's attributes, will there be any editor UI to control descendant styles from the parent block?

I'm imagining something like this in theme.json

```json
	"styles": {
		"spacing": {
			"margin": "44px"
		},
		"blocks": {
			"core/columns": {
				"spacing": {
					"padding": "22px"
				},
				"blocks": {
					"core/social-links": {
						"color": {
							"background": "kermit-green"
						}
					}
				}
			}
		}
	},
```

That would be the recipe for the following CSS:

```css
body {
    margin: 44px;
}
.wp-block-columns {
    padding: 22px;
}
.wp-block-columns .wp-block-social-links {
    background-color: kermit-green;
}
```

Sorry if I'm rehashing old ground and asking a lot of questions! Just trying to get context. 🙇